File: tests/benchmarks/tcp/echo_server.go
package main

import (
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
	"sync"
	"time"
)

func serveMessage(conn net.Conn) {
	defer conn.Close()
	buf := make([]byte, 65536)
	for {
		n, err := conn.Read(buf)
		if err != nil {
			if err != io.EOF {
				// read error, silently close
			}
			return
		}
		written := 0
		for written < n {
			w, err := conn.Write(buf[written:n])
			if err != nil {
				return
			}
			written += w
		}
	}
}

func serveDiscard(conn net.Conn) {
	defer conn.Close()
	buf := make([]byte, 65536)
	for {
		if _, err := conn.Read(buf); err != nil {
			if err == io.EOF {
				conn.Write([]byte("OK"))
			}
			return
		}
	}
}

func serveSlowDiscard(conn net.Conn, delayMs int64) {
	defer conn.Close()
	buf := make([]byte, 4096)
	delay := time.Duration(delayMs) * time.Millisecond
	for {
		if _, err := conn.Read(buf); err != nil {
			if err == io.EOF {
				conn.Write([]byte("OK"))
			}
			return
		}
		if delay > 0 {
			time.Sleep(delay)
		}
	}
}

func serveSource(conn net.Conn, totalBytes int64) {
	defer conn.Close()
	chunk := make([]byte, 65536)
	for i := range chunk {
		chunk[i] = 'S'
	}
	remaining := totalBytes
	for remaining > 0 {
		data := chunk
		if remaining < int64(len(chunk)) {
			data = chunk[:int(remaining)]
		}
		n, err := conn.Write(data)
		if err != nil || n <= 0 {
			return
		}
		remaining -= int64(n)
	}
}

func serveProxy(client net.Conn, upstreamPort int) {
	upstream, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%d", upstreamPort))
	if err != nil {
		client.Close()
		return
	}

	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		io.Copy(upstream, client)
		upstream.Close()
		client.Close()
	}()
	go func() {
		defer wg.Done()
		io.Copy(client, upstream)
		client.Close()
		upstream.Close()
	}()
	wg.Wait()
}

func handleConn(conn net.Conn, mode string, extra int64) {
	switch mode {
	case "stream", "message", "idle":
		serveMessage(conn)
	case "discard":
		serveDiscard(conn)
	case "slow_discard":
		serveSlowDiscard(conn, extra)
	case "source":
		serveSource(conn, extra)
	case "proxy":
		serveProxy(conn, int(extra))
	default:
		serveMessage(conn)
	}
}

func main() {
	port := 9001
	if len(os.Args) > 1 {
		if p, err := strconv.Atoi(os.Args[1]); err == nil {
			port = p
		}
	}
	mode := "message"
	if len(os.Args) > 2 {
		mode = os.Args[2]
	}
	extra := int64(64 * 1024 * 1024)
	if len(os.Args) > 3 {
		if v, err := strconv.ParseInt(os.Args[3], 10, 64); err == nil {
			extra = v
		}
	}

	addr := fmt.Sprintf(":%d", port)
	listener, err := net.Listen("tcp", addr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ERROR: %v\n", err)
		os.Exit(1)
	}
	defer listener.Close()

	fmt.Printf("Go TCP benchmark server listening on port %d mode=%s\n", port, mode)

	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		go handleConn(conn, mode, extra)
	}
}

File: tests/benchmarks/tcp/echo_server.js
// TCP benchmark server for Node.js
// Usage: node echo_server.js [port mode extra]

const net = require('net');

const port = parseInt(process.argv[2]) || 9001;
const mode = process.argv[3] || 'message';
const extra = parseInt(process.argv[4]) || (64 * 1024 * 1024);
const chunk = Buffer.alloc(65536, 'S');

function serveMessage(socket) {
    socket.on('data', (data) => {
        socket.write(data);
    });
    socket.on('error', () => {});
}

function serveDiscard(socket) {
    socket.on('data', () => {});
    socket.on('end', () => {
        socket.end('OK');
    });
    socket.on('error', () => {});
}

function serveSlowDiscard(socket, delayMs) {
    socket.on('data', () => {
        socket.pause();
        setTimeout(() => socket.resume(), Math.max(0, delayMs));
    });
    socket.on('end', () => {
        socket.end('OK');
    });
    socket.on('error', () => {});
}

function serveSource(socket, totalBytes) {
    let remaining = totalBytes;
    function writeMore() {
        while (remaining > 0) {
            const data = remaining >= chunk.length ? chunk : chunk.subarray(0, remaining);
            remaining -= data.length;
            if (!socket.write(data)) {
                socket.once('drain', writeMore);
                return;
            }
        }
        socket.end();
    }
    socket.on('error', () => {});
    writeMore();
}

function serveProxy(socket, upstreamPort) {
    const upstream = net.createConnection({ host: '127.0.0.1', port: upstreamPort });
    socket.pipe(upstream);
    upstream.pipe(socket);
    socket.on('error', () => upstream.destroy());
    upstream.on('error', () => socket.destroy());
}

const server = net.createServer((socket) => {
    if (mode === 'discard') {
        serveDiscard(socket);
    } else if (mode === 'slow_discard') {
        serveSlowDiscard(socket, extra);
    } else if (mode === 'source') {
        serveSource(socket, extra);
    } else if (mode === 'proxy') {
        serveProxy(socket, extra);
    } else {
        serveMessage(socket);
    }
});

server.listen(port, () => {
    console.log(`Node.js TCP benchmark server listening on port ${port} mode=${mode}`);
});
